File: Logging/ThresholdLogger.cs
using System;
using System.IO;

using System.Runtime.CompilerServices;

namespace Logging;

/// <summary>
/// This class is theoretically thread safe
/// </summary>
public class ThresholdLogger : ILogger
{
    private readonly LogLevel _printThreshold;
    private readonly LogLevel _storeThreshold;
    private uint _fileIndex = 1;
    private ReaderWriterLockSlim _rwLock = new ReaderWriterLockSlim();

    public ThresholdLogger() : this(LogLevel.Info, LogLevel.Info)
    {
        
    }
    
    public ThresholdLogger(LogLevel printThreshold, LogLevel storeThreshold)
    {
        this._printThreshold = printThreshold;
        this._storeThreshold = storeThreshold;
    }
    
    // TODO: callingClass here can be show sensitive information (filepaths), find out if this is a problem or not
    public void Log(LogLevel severityLevel, string logData, [CallerMemberName] string callingFunction = "", [CallerFilePath] string callingClass = "")
    {
        _rwLock.EnterWriteLock();
        try
        {
            HandleLogPrint(severityLevel, logData, callingClass, callingFunction);
            HandleStoringOfLog(severityLevel, logData, callingClass, callingFunction);
        }
        catch (Exception e)
        {
            Console.WriteLine($"Something failed asynchronously when logging. Error: {e}");
        }
        finally
        {
            _rwLock.ExitWriteLock();
        }
    }
    private void HandleLogPrint(LogLevel severityLevel, string logData, string callingClass, string callingFunction)
    {
        if (_printThreshold == LogLevel.Ignore || severityLevel < _printThreshold) return;
        
        Console.WriteLine(CreateLoggingMessage(severityLevel, logData, callingClass, callingFunction));
    }

    private string CreateLoggingMessage(LogLevel severityLevel, string logData, string callingClass, string callingFunction)
    {
        int lastBackslashPosition = callingClass.LastIndexOf("\\", StringComparison.Ordinal) + 1;
        string className = callingClass.Substring(lastBackslashPosition, callingClass.Length - lastBackslashPosition);
        return $"{DateTime.Now} [{severityLevel}] in {callingFunction} in {className} | {logData}";
    }
    
    private void HandleStoringOfLog(LogLevel severityLevel, string logData, string callingClass, string callingFunction)
    {
        if (_storeThreshold == LogLevel.Ignore || severityLevel < _storeThreshold) return;

        var filePath = CreateFilePath();

        try
        {
            StreamWriter writer = new StreamWriter(filePath, true);
            writer.WriteLine(CreateLoggingMessage(severityLevel, logData, callingClass, callingFunction));
            writer.Dispose();
        }
        catch (Exception e)
        {
            Console.WriteLine(CreateLoggingMessage(LogLevel.Error, $"Failed to store data to file. Error: {e}. Data" + CreateLoggingMessage(severityLevel, logData, callingClass, callingFunction),  callingClass, callingFunction));
        }
    }

    private string CreateFilePath()
    {
        var fileName = CreateFileName();
        var filePath = CreateFilePathForFileName(fileName);

        do
        {
            if (new FileInfo(filePath).Length < LoggingConstants.MaxFileSize) break;
            _fileIndex++;
            fileName = CreateFileName();
            filePath = CreateFilePathForFileName(fileName);
        } while (File.Exists(filePath));

        return filePath;
    }

    private string CreateFileName()
    {
        return $"threshold_logger_{DateTime.Now.Date:d}_{_fileIndex}.txt";
    }

    private string CreateFilePathForFileName(string fileName)
    {
        var folderPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, LoggingConstants.FolderName);
        return Path.Combine(folderPath, fileName);
    }
    
}

File: logging/src/threshold_logger.rs
use chrono::Local;

use crate::logger::{LogData, LogLevel, Logger};
use crate::{LOG_FOLDER_NAME, MAX_FILE_SIZE};
use std::any::type_name;
use std::env;
use std::fs::{metadata, OpenOptions};
use std::io::Write;
use std::path::Path;

pub struct ThresholdLogger {
    print_threshold: LogLevel,
    store_threshold: LogLevel,
    file_index: u128,
}

impl ThresholdLogger {
    #[must_use]
    pub const fn new(print_threshold: LogLevel, store_threshold: LogLevel) -> Self {
        Self {
            print_threshold,
            store_threshold,
            file_index: 0,
        }
    }

    fn handle_log_print(&mut self, data: LogData) {
        if self.print_threshold == LogLevel::Ignore || data.severity_level < self.print_threshold {
            return;
        }

        println!("{data}");
    }

    fn handle_storing_of_log(&mut self, data: LogData) {
        if self.store_threshold == LogLevel::Ignore || data.severity_level < self.store_threshold {
            return;
        }

        let file_path: String = match self.create_file_path() {
            Ok(path) => path,
            Err(e) => {
                let error_string = format!("Failed to make get filepath because: {e}");
                let write_log =
                    LogData::new(LogLevel::Error, error_string.as_str(), type_name::<Self>());
                self.handle_log_print(write_log);
                return;
            }
        };

        match OpenOptions::new().append(true).create(true).open(file_path) {
            Ok(file) => self.write_to_file(file, data),
            Err(e) => {
                let error_string = format!("Failed to open file because: {e}");
                let write_log =
                    LogData::new(LogLevel::Error, error_string.as_str(), type_name::<Self>());
                self.handle_log_print(write_log);
            }
        }
    }

    fn write_to_file(&mut self, mut file: std::fs::File, data: LogData) {
        match writeln!(file, "{data}") {
            Ok(_) => (),
            Err(e) => {
                let error_string = format!("Failed to write {data} to file. Error: {e}");
                let write_log =
                    LogData::new(LogLevel::Error, error_string.as_str(), type_name::<Self>());
                self.handle_log_print(write_log);
            }
        }
    }

    fn create_file_path(&mut self) -> Result<String, String> {
        let mut file_name: String = self.create_file_name();
        let mut file_path: String = match Self::create_file_path_for_file_name(&file_name) {
            Ok(path) => path,
            Err(e) => return Err(format!("Failed to create file path because: {e}")),
        };

        while metadata(&file_path).map(|m| m.len()).unwrap_or(0) >= MAX_FILE_SIZE {
            self.file_index += 1;
            file_name = self.create_file_name();
            file_path = match Self::create_file_path_for_file_name(&file_name) {
                Ok(path) => path,
                Err(e) => return Err(format!("Failed to create file path because: {e}")),
            };
        }

        Ok(file_path)
    }

    fn create_file_name(&self) -> String {
        format!(
            "threshold_logger_{}_{}.txt",
            Local::now().format("%d-%m-%Y"),
            self.file_index
        )
    }

    fn create_file_path_for_file_name(file_name: &str) -> Result<String, String> {
        match env::current_exe() {
            Ok(path) => match path.parent() {
                Some(exe_folder) => {
                    let file_path =
                        Path::new(exe_folder).join(LOG_FOLDER_NAME.to_string() + "/" + file_name);
                    return Ok(file_path.to_string_lossy().to_string());
                }
                None => Err("Failed to get path of the folder the executable is in.".to_string()),
            },
            Err(_) => Err("Failed to get the path to the executable.".to_string()),
        }
    }
}

impl Logger for ThresholdLogger {
    fn log(&mut self, data: LogData) {
        self.handle_log_print(data);
        self.handle_storing_of_log(data);
    }
}
